### DISCUSSION TITLE
chapter_array_and_linkedlist/summary/

### DISCUSSION THREAD

**Discussion by giscus** [krahets/hello-algo#150] [Announcements]
# chapter_array_and_linkedlist/summary/

一本动画图解、能运行、可提问的数据结构与算法入门书

https://www.hello-algo.com/chapter_array_and_linkedlist/summary/

**Comment by fangichao** [comment]
真棒！！！

**Comment by HandyCC** [comment]
最后一个Question的 `res = [[0] * self.size()]` 是否写错了，应该是`res = [[0]] * self.size()`?

**Reply by krahets** [reply]
谢谢指正！

**Comment by DayDayYES** [comment]
打卡，数组和链表两种线性数据结构，都支持数据插入，删除，查询，修改.数组空间长度不可变，空间效率低.链表分散存储的特性使得它可以有效利用空间，但链表的查修操作的时间复杂度为O（n）.列表是一种抽象的数据结构概念，可以由数组或者链表形成，它也称为动态数组，可以动态地进行扩容，从而克服了数组长度不可变的缺点

**Comment by HP-Patience** [comment]
2024.5.29直接拿下打卡！

数组和链表这两种数据结构是互补的，
数组对于元素访问优于链表，
链表对于元素删除和增加优于数组。

今天才知道列表原来是一种抽象的数据结构概念，是可以动态修改并支持增删查改的数据结构，之前一直和链表混着用。

还有这两种数据结构对于缓存机制有着不同的效益，综合看用数组偏多，实际还是看具体需求，链表更多地应用在数据动态性更强的案例中。

发现不管是力扣题目还是hello算法介绍，在python语言上，都会去使用类别注释，个人觉得还是比较好的习惯，我之前觉得麻烦一直没有使用类别注释，觉得后面需要培养这个习惯。因为python天生支持多态的特性，而其他语言比如java和c++都需要明确指出数据类型，对于代码解释性更强。

**Comment by shaoxinH** [comment]
列表这里可以看一下Java的源码 虽然我算法学的是c++ 哈哈。。使用c++实时模拟一下列表实现，很有用，特别是感受数组扩容与设置报错的过程，对自己的编程逻辑很有好处

**Comment by tiam-bloom** [comment]
每一章或者每一小节后面是不是可以给一些相关的算法题(或者力扣题的地址)以供练习更好

**Reply by a2635814919** [reply]
这个作者的git库[LeetCode-Book](https://github.com/krahets/LeetCode-Book)，里面有给出推荐的算法题

**Reply by AddictedK1** [reply]
How can u type this??

**Comment by HuaWang135608** [comment]
链表增删的时候没有元素移动



---- 回复的原邮件 ----
| 发件人 | ***@***.***> |
| 日期 | 2024年09月09日 15:42 |
| 收件人 | ***@***.***> |
| 抄送至 | ***@***.***>***@***.***> |
| 主题 | Re: [krahets/hello-algo] chapter_array_and_linkedlist/summary/ (Discussion #150) |

我有一个疑问，既然链表访问的时间复杂度为O(n)，那链表想要增删某一个元素时不是要先访问再增删吗，时间复杂度为O(n);数组增删的时间复杂度为O(n),为什么说链表在增删上效率更高

**Comment by Jeffstarm** [comment]
"栈上的数组的大小需要在编译时确定，而堆上的数组的大小可以在运行时动态确定"，哪位大哥可以解释下这句话的意思

**Reply by SDULinW** [reply]
栈内存分配是静态的，也就是说，在编译阶段需要确定数组的大小。例如：
```cpp
int arr[10]; // 创建一个大小为 10 的数组
```
在这个例子中，数组的大小 `10` 是固定的，编译器会在编译时为该数组分配足够的栈内存。

堆内存分配是动态的，数组的大小可以在运行时根据程序的需要来决定。例如：
```cpp
int n;
std::cin >> n; // 用户输入数组的大小
int* arr = new int[n]; // 在堆上创建一个大小为 n 的数组
```
在这个例子中，数组的大小 `n` 是在运行时通过用户输入决定的，编译器无法在编译阶段确定这个大小。

**Comment by shyGuide** [comment]
> Q：在链表中插入和删除操作的时间复杂度是 
 。但是增删之前都需要 
 的时间查找元素，那为什么时间复杂度不是 
 呢？
如果是先查找元素、再删除元素，时间复杂度确实是 
 。然而，链表的 
 增删的优势可以在其他应用上得到体现。例如，双向队列适合使用链表实现，我们维护一个指针变量始终指向头节点、尾节点，每次插入与删除操作都是 
 。

这个还是没太理解，有没有大神讲解一下

**Reply by Pgq-Selia** [reply]
此处的插入和删除操作单单只这两个操作本身
默认已知需要插入/删除的位置，操作本身的时间复杂度是O（1）

如果不知道需要插入/删除的位置，则需要先查找，查找耗时最坏情况是O（n）
那么整个操作的时间复杂度就变成O（n）了

**Reply by JoneKong** [reply]
我也没理解，直到问了AI：
你提到的问题实际上涉及到链表操作的时间复杂度分析中的两个不同方面：插入/删除操作本身的时间复杂度，以及查找元素的时间复杂度。

插入和删除操作的时间复杂度：
在链表中，如果已经找到了要插入或删除的节点（或者其前驱节点），那么插入和删除操作的时间复杂度确实是 O(1)。这是因为插入和删除只需要调整指针，不需要移动其他节点。
查找元素的时间复杂度：
查找一个特定元素的时间复杂度是 O(n)，因为在最坏情况下，你可能需要遍历整个链表才能找到目标元素。这适用于顺序查找的情况，即没有额外索引或其他加速查找机制的情况下。
因此，如果你需要在链表中插入或删除一个特定元素，并且这个元素不是通过某种方式预先知道位置的（例如，通过索引或哈希表辅助查找），那么总的时间复杂度确实会是 O(n)，因为查找占主导地位。

但是，如果我们讨论的是在已知位置进行插入或删除操作（例如，在头节点、尾节点或某个已知节点之后插入新节点），那么这些操作的时间复杂度仍然是 O(1)。

总结来说：

已知位置的插入和删除操作：O(1)
查找元素并在此基础上进行插入或删除：O(n)
因此，时间复杂度是否为 O(n) 取决于你具体的操作场景。如果查找是必须的步骤，那么整体复杂度就是 O(n)；否则，对于已知位置的操作，复杂度是 O(1)。

**Comment by ElvisWai** [comment]
Q：操作 res = [[0]] * n 生成了一个二维列表，其中每一个 [0] 都是独立的吗？
其实有歧义，应该是 res[0][n] 的每个值 `0` 是否独立？
```python
res = [[0]] * 3  # [[0], [0], [0]]
res[0][0] = 1 # [[1], [1], [1]]

res = [[0] * 3]  #  [[0], [0], [0]]
res[0]= [1] # [[1], [0], [0]]
```

**Reply by yhh2908307500** [reply]
其实也没有歧义。“其中每一个 [0] 都是独立的吗？” 这句话包含了 “其中每一个 [0] 中的0 都是独立的吗？”

**Comment by NotUr77** [comment]
“Python 中的数字也被包装为对象，列表中存储的不是数字本身，而是对数字的引用” 长见识了

**Comment by jiuerya7** [comment]
好家伙，python里面变量存储的都是对象的引用，涨知识了，那是不是所有弱类型语言的实现都是和python一样？

**Comment by xiaozhuge123** [comment]
day03

**Comment by RobinElysia** [comment]
在该列表中，所有整数 0 都是同一个对象的引用。这是因为 Python 对小整数（通常是 -5 到 256）采用了缓存池机制，以便最大化对象复用，从而提升性能。

虽然它们指向同一个对象，但我们仍然可以独立修改列表中的每个元素，这是因为 Python 的整数是“不可变对象”。当我们修改某个元素时，实际上是切换为另一个对象的引用，而不是改变原有对象本身。

然而，当列表元素是“可变对象”时（例如列表、字典或类实例等），修改某个元素会直接改变该对象本身，所有引用该对象的元素都会产生相同变化。

**Reply by RobinElysia** [reply]
这点涉及到python的深浅拷贝了，感兴趣的可以查一下

**Comment by solink7** [comment]
# 打卡2025/6/8
太精彩了！

**Comment by Ivy129** [comment]
2025年6月12日
打卡，这段总结仍然还有存在疑惑的地方

**Comment by main520** [comment]
##### 打卡 2025/7/7

###### 好困

**Comment by zhuchentong** [comment]
坚持学习，打卡20250805

**Comment by nanmingjushi** [comment]
在Java中，ArrayList 和 LinkedList 都是 Java 中 List 接口的实现类，ArrayList 基于动态数组，LinkedList 基于双向链表。

**Comment by Oceans820** [comment]
有点不是很理解列表和链表直接的区别，希望有友友可以帮忙解答，感谢

**Reply by KuitoInoguchi** [reply]
列表是一种更抽象的数据结构，而链表是比较具体的数据结构，因此列表可以由链表实现，也可以由数组实现。你可以把列表想象成“交通工具”，而链表是“飞机”，数组是“汽车”，“飞机”或者“汽车”都可以承担“交通工具”的职责，只是各有各的优势和短板。

**Reply by Ke-Albert** [reply]
可不可以这么认为，数组和链表都是贴近计算机底层的数据结构，偏向于和内存等这些硬件设备打交道，它们两个更现实；而基于数组和链表衍生出来的其它数据结构，比如列表，则是在此基础之上的进一步抽象，这些数据结构隐藏了一些底层的实现细节。

**Comment by zhaoyinchen84-svg** [comment]
10月20日 打卡！

**Comment by wink0521** [comment]
栈上的数组的大小需要在编译时确定，而堆上的数组的大小可以在运行时动态确定。
所以 java的方法的局部变量必须赋默认值，而成员变量可以不给默认值。

**Comment by alonewolf4068** [comment]
结论中的每一项，希望有个代码示例，否则对算法不是很熟练的同学感觉很抽象， 感谢

**Reply by alonewolf4068** [reply]
比如最后两项，感觉有点世懵

**Comment by ZhyBing** [comment]
Q：在链表中插入和删除操作的时间复杂度是 O(1)这个结论简单来说指的应该是这两个操作本身，不包括查找的过程，前置条件设为已经知道了要插入的位置，就和段末所说双向链表的插入和删除这种应用上，理论上是已经给定了位置，仅观察插入或删除本身。

**Comment by believebelief** [comment]
问题处有些看不明白
